### PR TITLE
8281764: jextract does not generate parameter names for function pointer typedefs

### PR DESCRIPTION
**Pull request by sundararajana**
Propagating TypedefDecl's ParmDecl child cursors to Types.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8281764](https://bugs.openjdk.java.net/browse/JDK-8281764): jextract does not generate parameter names for function pointer typedefs

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - Committer) ⚠️ Review applies to 0a8d2e20cd3782d9287018cc2f0a5dbd1cf26c7a

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/646/head:pull/646` \
`$ git checkout pull/646`

Update a local copy of the PR: \
`$ git checkout pull/646` \
`$ git pull https://git.openjdk.java.net/panama-foreign pull/646/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 646`

View PR using the GUI difftool: \
`$ git pr show -t 646`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-foreign/pull/646.diff">https://git.openjdk.java.net/panama-foreign/pull/646.diff</a>

</details>

### CODE REVIEW

File: src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/TypeMaker.java [hunk 1]
```diff
@@ -118,7 +124,7 @@ Type makeType(jdk.internal.clang.Type t) {
         }
     }
 
-    Type makeTypeInternal(jdk.internal.clang.Type t) {
+    Type makeTypeInternal(jdk.internal.clang.Type t, List<Declaration.Variable> params) {
```
Comment (mcimadamore) [inline src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/TypeMaker.java:127]: I believe you can avoid all the changes to this method, and have a simple method on FunctionType which gets a parameter name list and gives you a _new_ function type with the list in it - e.g. `withParameterNames`. Then you can call that method from `TreeMaker` _after_ you have parsed a `Type`.

**Review by mcimadamore — APPROVED** [review]
Looks good - I've added a suggestion to perhaps reduce the impact of the changes on TypeMaker.

File: src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/TreeMaker.java [hunk 2]
```diff
@@ -257,6 +257,37 @@ private static boolean isAnonymousStruct(Declaration declaration) {
                 return null;
             }
         }
+        Type.Function funcType = null;
+        boolean isFuncPtrType = false;
+        if (canonicalType instanceof Type.Function) {
```
Comment (mcimadamore) [inline src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/TreeMaker.java:262]: I believe `OutputFactory::getAsFunctionPointer` as code that is similar to this (given a type, return it as a function type - following typedefs if necessary). Maybe we should move this into some util function and share?

**Review by mcimadamore — APPROVED** [review]

File: src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/TreeMaker.java (hunk 2 above)
Comment (sundararajana) [reply src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/TreeMaker.java:262]: That part of the code assumes all Types are properly resolved. In TreeMaker, we're in the middle TypeMaker's constructor (resolution happens later).

Comment (sundararajana) [reply src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/TreeMaker.java:262]: refactored pointer type check to Utils.java.

**Pull request closed without merging**